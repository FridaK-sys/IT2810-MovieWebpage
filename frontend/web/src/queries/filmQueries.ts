import { gql } from '@apollo/client'


/** 
* Query for searing for films
*/
export const SEARCH_FILMS = gql`
  query getPosts($limit: Int!, $offset: Int!, $titleFilter: String, $genreFilter: String, $yearFilter: Int, $sorting: Int) {
    getFilteredPosts(limit: $limit, offset: $offset, titleFilter: $titleFilter, genreFilter: $genreFilter, yearFilter: $yearFilter, sorting: $sorting) {
        _id
        title
        year
        cast
        genres
    }
  },
`;
/** 
* Query for adding a film
*/
export const ADD_FILM = gql`
  mutation createPost($title: String!, $year: Int, $cast: [String], $genres: [String]) {
    createPost(title: $title, year: $year, cast: $cast, genres: $genres) {
      _id
      title
      year
      cast
      genres
    }
  }
`