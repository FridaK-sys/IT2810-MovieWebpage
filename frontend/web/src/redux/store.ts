import {applyMiddleware} from "redux";
import { legacy_createStore as createStore} from 'redux'

import thunk from "redux-thunk";

export interface Store {
    title: string;
    genre: string;
    year: string;
    sorting: string;
}

// Redux implementation
function storeReducer(state: Store = {
    title: "",
    genre: "",
    year: "0",
    sorting: "1",
}, action: any) {
    switch (action.type) {
        case "SET_TITLE":
            return {
                ...state,
                title: action.payload
            };
        case "SET_GENRE":
            return {
                ...state,
                genre: action.payload
            };
        case "SET_YEAR":
            return {
                ...state,
                year: action.payload
            };
        case "SET_SORTING":
            return {
                ...state,
                sorting: action.payload
            };
        default:
            return state;
    }
}

// apply thunk middleware to store
const store = createStore(storeReducer,applyMiddleware(thunk));

export default store;