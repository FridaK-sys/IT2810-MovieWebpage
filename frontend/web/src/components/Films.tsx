import { useMutation, useQuery } from '@apollo/client'
import { Button, DatePicker, DatePickerProps, Select, Col, Card, Row } from 'antd';
import Search from 'antd/lib/input/Search';
import moment from 'moment';
import { useState } from 'react';
import { optionList } from '../helpers/helpers';
import { SEARCH_FILMS, ADD_FILM } from '../queries/filmQueries';
import { Film } from '../utils/Interface';
import { CollectionCreateForm } from './AddFilm';
import { ShowFilmItem } from './FilmItem';

const PAGE_SIZE = 15;

export default function Films() {
    const [page, setPage] = useState(0);
    const [filterInput, setFilterInput] = useState<String>("");
    const [titleFilter, setTitleFilter] = useState<String>("");
    const [genreFilter, setGenreFilter] = useState<String>("");
    const [yearFilter, setYearFilter] = useState(0);
    const [openCreate, setOpenCreate] = useState(false);
    const [currentPost, setCurrentPost] = useState<Film>({
        _id: "",
        title: "",
        year: "",
        cast: [""],
        genres: [""]
    });

    const [isModalOpen, setIsModalOpen] = useState(false);

    const handleCancel = () => {
      setIsModalOpen(false);
    };

    const [createPost] = useMutation(ADD_FILM);
    
    const { loading, error, data } = useQuery(SEARCH_FILMS, {
        variables: {
            limit: PAGE_SIZE,
            offset: page * PAGE_SIZE,
            titleFilter: titleFilter,
            genreFilter: genreFilter,
            yearFilter: yearFilter,
        },
    });

    function disabledYear(current: any) {
        return current.year() > 2022 || current.year() < 1900;
    }

    if (loading) {
        return (
            <div className='container'>
                <div className='d-flex justify-content-center mt-3'>
                    <div className='spinner-border' role='status'>
                        <span className='sr-only'></span>
                    </div>
                </div>
            </div>
        )
    }

   if (error) {
        console.log(error)
        return (
            <div className='container'>
                <div className='mt-3'>
                    <p>Something went wrong...</p>
                </div>
            </div>
        )
    }
    
    const onCreate = (values: any) => {
        setYearFilter(0);
        setGenreFilter("");
        setTitleFilter("");

        createPost({
            variables: {
                title: values.title,
                year: values.year? parseInt(values.year, 10) : null,
                cast: values.cast? values.cast.split(",") : null,
                genres: values.genres? [values.genres]: null,
            }
        });
        setTitleFilter(values.title);
        setOpenCreate(false);
    };

    const handleFilterInput = (input: string) => {
        setFilterInput(input);
    };

    const changeTitle = () => {
        setTitleFilter(filterInput);
        setGenreFilter("");
        setYearFilter(0);
    };

    const changeGenre = (value: string) => {
        setGenreFilter(value);
        setTitleFilter("");
        setYearFilter(0);
    };

    const changeDate: DatePickerProps['onChange'] = (date, dateString) => {
        setYearFilter(parseInt(dateString, 10))
        setTitleFilter("");
        setGenreFilter("");
    };

    function handleClick(post: Film) {
        console.log(post.title)
        setCurrentPost(post);
        setIsModalOpen(true);
    }

    let body: any = [];
    body.push (
        data.getFilteredPosts?.map((post: Film) => (
            <Col xs={24} md={6} className="my-4 mx-2">
                <Card hoverable={true} onClick={() => handleClick(post)} title={post.title} style={{textAlign: "center"}} >
                    <p style={{textAlign: "center"}}> Year Released: {post.year? post.year: ""} </p>
                </Card>
            </Col>
        ))
    )

    return (
        <>
        {!loading && !error && 
            <div className='container m-3 pb-3'>    
                <div className='d-flex pt-2' style={{justifyContent: "center"}}>
                    <div className='px-2'>
                            <Search defaultValue={titleFilter !== ""? titleFilter.toString(): undefined} placeholder="Search for title" onChange={(e) => handleFilterInput(e.target.value)} onSearch={changeTitle} style={{ width: 400 }} />
                        </div>
                        <div className='px-2'>
                            <Select defaultValue={genreFilter !== ""? genreFilter.toString(): undefined} style={{ width: 200 }} placeholder="Choose a genre" onChange={changeGenre}>
                                {optionList}
                            </Select>
                        </div>
                        <div className='px-2'>
                            <DatePicker disabledDate={disabledYear} defaultValue={(yearFilter !== 0)? moment(yearFilter.toString()) : undefined} style={{ width: 200 }} placeholder="Choose a year" onChange={changeDate} picker="year" />
                        </div>
                        <div className='px-2'>
                            <Button
                                type="primary"
                                onClick={() => {
                                    setOpenCreate(true);
                                }}
                            >
                                New Film
                            </Button>
                            <CollectionCreateForm
                                open={openCreate}
                                onCreate={onCreate}
                                onCancel={() => {
                                    setOpenCreate(false);
                                }}
                            /> 
                        </div>
                    </div>

                    <div className="site-card-wrapper">
                        <Row justify="center" align='middle'>
                            {body}
                        </Row>
                        <ShowFilmItem 
                            film={currentPost} 
                            open={isModalOpen} 
                            onCancel={handleCancel} 
                        />
                    </div>

                    <div className='mt-2 mb-2' style={{textAlign: "center"}}>
                        <Button
                            type='primary'
                            className="m-2"
                            id="buttonLoadMore"
                            disabled={loading}
                            onClick={() => (setPage(prev => prev-1))}
                        >
                            Previous
                        </Button>
                        <Button
                            type='primary'
                            id="buttonLoadMore"
                            disabled={loading}
                            onClick={() => (setPage(prev => prev+1))}
                        >
                            Next
                        </Button>
                    </div>
                </div>
            }
        </>
    )
}