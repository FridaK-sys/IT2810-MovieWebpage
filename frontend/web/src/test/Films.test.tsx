import { fireEvent, getByTestId, queryByText, render, waitFor,screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect'
import '@testing-library/jest-dom'
import Films from '../components/Films';
import { MockedProvider} from "@apollo/client/testing";
import { SEARCH_FILMS } from '../queries/filmQueries';
import { Provider } from 'react-redux';
import store from '../redux/store';
import userEvent from '@testing-library/user-event';


const mocks=  [
  {
    request: {
      query: SEARCH_FILMS,
      variables: {
        limit: 15,
        offset:"" ,
        titleFilter: "",
        genreFilter: "",
        yearFilter: 0,
        sorting: 1

      }
    },
    result: {
      data: {
        getFilteredPosts: [{"_id":"63453c374d8655244d2aa472","title":"After Dark in Central Park","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa473","title":"Boarding School Girls' Pajama Parade","year":1900,"cast":[],"genres":[]},
        {"_id":"63453c374d8655244d2aa474","title":"Buffalo Bill's Wild West Parad","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa475","title":"Caught","year":1900,"cast":[],"genres":[]},
        {"_id":"63453c374d8655244d2aa476","title":"Clowns Spinning Hats","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa477","title":"Capture of Boer Battery by British","year":1900,"cast":[],"genres":["Short","Documentary"]},
        {"_id":"63453c374d8655244d2aa478","title":"The Enchanted Drawing","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa479","title":"Feeding Sea Lions","year":1900,"cast":["Paul Boyton"],"genres":[]},{"_id":"63453c374d8655244d2aa47b","title":"New Life Rescue","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa47c","title":"New Morning Bath","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa47d","title":"Searching Ruins on Broadway, Galveston, for Dead Bodies","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa47e","title":"The Tribulations of an Amateur Photographer","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa47f","title":"Trouble in Hogan's Alley","year":1900,"cast":[],"genres":["Comedy"]},{"_id":"63453c374d8655244d2aa480","title":"Two Old Sparks","year":1900,"cast":[],"genres":["Short"]},{"_id":"63453c374d8655244d2aa481","title":"The Wonder, Ching Ling Foo","year":1900,"cast":["Ching Ling Foo"],"genres":["Short"]}]
      }
    }
  }
];

it("renders without error", async () => {
  render(
    <Provider store = {store}>
    <MockedProvider mocks={mocks} addTypename={false}>
      <Films />
    </MockedProvider>
    </Provider>
  );
  expect(await screen.findByText('Loading...')).toBeInTheDocument();
  
});

it("should show error UI", async () => {
  const filmMock = {
    request: {
      query: SEARCH_FILMS,
      variables: { 
        limit: 15,
        offset:0 ,
        titleFilter: "",
        genreFilter: "",
        yearFilter: 0,
        sorting: 1
       }
    },
    error: new Error("Something went wrong...")
  };
  render(
    <Provider store={store}>
    <MockedProvider mocks={[filmMock]} addTypename={false}>
      <Films />
    </MockedProvider>
    </Provider>
  );
  expect(await screen.findByText("Something went wrong...")).toBeInTheDocument();
});

it("should render films", async () => {
  const filmMock = {
    request: {
      query: SEARCH_FILMS,
      variables: {
        limit: 15,
        offset:0 ,
        titleFilter: "",
        genreFilter: "",
        yearFilter: 0,
        sorting: 1

      }
    },
    result: {
      data: {
        getFilteredPosts: [{"_id":"63453c374d8655244d2aa472","title":"After Dark in Central Park","year":1900,"cast":[],"genres":[]},
        {"_id":"63453c374d8655244d2aa473","title":"Boarding School Girls' Pajama Parade","year":1900,"cast":[],"genres":[]},
        {"_id":"63453c374d8655244d2aa474","title":"Buffalo Bill's Wild West Parad","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa475","title":"Caught","year":1900,"cast":[],"genres":[]},
        {"_id":"63453c374d8655244d2aa476","title":"Clowns Spinning Hats","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa477","title":"Capture of Boer Battery by British","year":1900,"cast":[],"genres":["Short","Documentary"]},
        {"_id":"63453c374d8655244d2aa478","title":"The Enchanted Drawing","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa479","title":"Feeding Sea Lions","year":1900,"cast":["Paul Boyton"],"genres":[]},{"_id":"63453c374d8655244d2aa47b","title":"New Life Rescue","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa47c","title":"New Morning Bath","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa47d","title":"Searching Ruins on Broadway, Galveston, for Dead Bodies","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa47e","title":"The Tribulations of an Amateur Photographer","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa47f","title":"Trouble in Hogan's Alley","year":1900,"cast":[],"genres":["Comedy"]},{"_id":"63453c374d8655244d2aa480","title":"Two Old Sparks","year":1900,"cast":[],"genres":["Short"]},{"_id":"63453c374d8655244d2aa481","title":"The Wonder, Ching Ling Foo","year":1900,"cast":["Ching Ling Foo"],"genres":["Short"]}]
      }
    }
  }
  render(
    <Provider store={store}>
    <MockedProvider mocks={[filmMock]} addTypename={false}>
      <Films  />
    </MockedProvider>
    </Provider>
  );
  expect(await screen.findByText("After Dark in Central Park")).toBeInTheDocument();
});

it("clear search field", async () => {
  const searchMock=  [
    {
      request: {
        query: SEARCH_FILMS,
        variables: {
          limit: 15,
          offset:"" ,
          titleFilter: "After Dark in Central Park",
          genreFilter: "",
          yearFilter: 0,
          sorting: 1
  
        }
      },
      result: {
        data: {
          getFilteredPosts: [{"_id":"63453c374d8655244d2aa472","title":"After Dark in Central Park","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa473","title":"Boarding School Girls' Pajama Parade","year":1900,"cast":[],"genres":[]},
          {"_id":"63453c374d8655244d2aa474","title":"Buffalo Bill's Wild West Parad","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa475","title":"Caught","year":1900,"cast":[],"genres":[]},
          {"_id":"63453c374d8655244d2aa476","title":"Clowns Spinning Hats","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa477","title":"Capture of Boer Battery by British","year":1900,"cast":[],"genres":["Short","Documentary"]},
          {"_id":"63453c374d8655244d2aa478","title":"The Enchanted Drawing","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa479","title":"Feeding Sea Lions","year":1900,"cast":["Paul Boyton"],"genres":[]},{"_id":"63453c374d8655244d2aa47b","title":"New Life Rescue","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa47c","title":"New Morning Bath","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa47d","title":"Searching Ruins on Broadway, Galveston, for Dead Bodies","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa47e","title":"The Tribulations of an Amateur Photographer","year":1900,"cast":[],"genres":[]},{"_id":"63453c374d8655244d2aa47f","title":"Trouble in Hogan's Alley","year":1900,"cast":[],"genres":["Comedy"]},{"_id":"63453c374d8655244d2aa480","title":"Two Old Sparks","year":1900,"cast":[],"genres":["Short"]},{"_id":"63453c374d8655244d2aa481","title":"The Wonder, Ching Ling Foo","year":1900,"cast":["Ching Ling Foo"],"genres":["Short"]}]
        }
      }
    }
  ];

  render(
    <Provider store={store}>
    <MockedProvider mocks={searchMock} addTypename={false}>
      <Films />
    </MockedProvider>
    </Provider>
  );


  // Find the button element...

  const button = await screen.getByRole('button', {
    name: /edit/
  })
  expect(button).toBeInTheDocument();

  // const searchField = await screen.findByAltText("Search for title") as HTMLInputElement;
  // userEvent.click(button); // Simulate a click and fire the mutation

 // const button2 = container.querySelector('reset')! as HTMLElement;
  //const searchField2 = container.querySelector('searchField') as HTMLInputElement;
 // expect(await screen.findByText("Buffalo Bill's Wild West Parad")).toBeInTheDocument();  
  //expect(searchField.value).toBe('');




});




