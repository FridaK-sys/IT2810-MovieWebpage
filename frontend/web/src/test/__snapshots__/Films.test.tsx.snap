// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jest Snapshot testing suite Matches DOM Snapshot 1`] = `
<div
  className="container d-flex justify-content-center mt-3"
>
  <div
    className="spinner-border"
    role="status"
  >
    <span
      className="sr-only"
    />
  </div>
  <p>
    Loading...
  </p>
</div>
`;
